### DISCUSSION TITLE
A gem5 Orchestration Framework (Geof!)

### DISCUSSION THREAD

**Discussion by BobbyRBruce** [gem5/gem5#754] [gem5-dev]
At UC Davis we've had this idea floating around for some time. The general high-level idea is a gem5 Orchestration Framework in which users can manage and monitor several running gem5 processes. We find managing many gem5 processes to be a particular pain-point for many users. For example, we find users will typically write a bash script to spin up large numbers of gem5 instances, each running a slightly different configuration, outputting to a slightly different directory. Even if this all goes well, it's a headache to keep track of the data (e.g., which run corresponds to which output, what is complete and what is still running, what needs rerun, etc.). Beyond this it's not uncommon for gem5 runs to fail and need rerun or, worse still, stall or crash silently with little information given to the users. At the time of writing there are no good ways to _inspect_ a gem5 process beyond inspecting the output, and no way to easily restart from a failed or interrupted state.

We wish to develop a framework to help users manage multiple gem5 instances and overcome some of these pain-point. This orchestration framework could encompass many useful features of varying levels of engineering difficulty. The reason for this discussion is to gather information and see what the most impactful features would be.

Off the top of my head, these features have been discussed before and could be incorporated into such a framework:

* Tooling to inspect the state of current gem5 processes.
    * Current ticks-per-second.
    * Live instructions-per-second.
    * Given the above 2, we can determine the gem5 simulation has stalled.
* gem5 updated to allow for pausing of run, or recovery from crash.
    * gem5 would periodically "save" its state and allow users to restore to that state if gem5 terminates unexpectedly. (I've been told Python's Pickle library would be useful for this).
    * The same functionality can be used to pause a gem5 run and continue at a later time.
* Automatic spawning of multiple gem5 instances from a single benchmark suite (i.e., pass the suite ID to a configuration scripts and have, automatically, each application in that suite run in parallel, with the orchestration framework automatically managing the processes and ensuring their outputs and stored in a structured manner).
* More structured and organized output of gem5 run data (basically a resurrection of the gem5Art project).
    * A clear and direct mapping between a gem5 simulation configuraiton and the output data.
    * Via the orchestration framework, clearly access the simulation configuration and the data being output, even while running.

**Comment by powerjg** [comment]
For point 2, pausing and recovering from a crash, I suggest using [CRIU](https://criu.org/Main_Page).

**Comment by powerjg** [comment]
For running things in parallel, here's a mock up of a potential solution: https://github.com/gem5/gem5/compare/develop...powerjg:gem5:jason/better-parallel
Branch: https://github.com/powerjg/gem5/tree/jason/better-parallel

**Comment by Sir-NoChill** [comment]
Since the original post, implementation of multisim is now mainline. I think that a lot of this functionality can be implemented on top of multisim, and I would like to look into crafting some of those integrations. I would like to focus on the last two features:
1. Spawning multiple gem5 instances from the same benchmark suite (kind of already there)
2. Better organization and structure for run data

My goal would be to start at multisim and then gather information from the child processes (likely through piclking) and aggregating it into a more usable format.

In terms of the original goal, was there more interest in a dashboard-like system to keep track of simulations in realtime, or was terminal output still an acceptable means of monitoring?
